src/Internal/DependencyManagement/AbstractServiceProvider.php

```diff
@@ -0,0 +1,139 @@
+	 *
+	 * @return DefinitionInterface The generated container definition.
+	 *
+	 * @throws ContainerException Error when reflecting the class, or class constructor is not public, or an argument has no valid type hint.
+	 */
+	protected function add_with_auto_arguments( string $class_name, $concrete = null, bool $shared = false ) : DefinitionInterface {
+		$definition = new Definition( $class_name, $concrete );
+
+		$constructor = $this->verify_class_and_concrete( $class_name, $concrete );
+
+		if ( ! is_null( $constructor ) ) {
+			$constructor_arguments = $constructor->getParameters();
+			foreach ( $constructor_arguments as $argument ) {
+				if ( $argument->isDefaultValueAvailable() ) {
+					$default_value = $argument->getDefaultValue();
+					$definition->addArgument( new RawArgument( $default_value ) );
+				} else {
+					$argument_class = $argument->getClass();
+					if ( is_null( $argument_class ) ) {
+						throw new ContainerException( "AbstractServiceProvider::add_with_auto_arguments: constructor argument '{$argument->getName()}' of class '$class_name' doesn't have a type hint or has one that doesn't specify a class." );
+					}
+
+					$definition->addArgument( $argument_class->name );
+				}
+			}
+		}
+
+		// Register the definition only after being sure that no exception will be thrown.
+
+		$this->getContainer()->add( $definition->getAlias(), $definition, $shared );
+
+		return $definition;
+	}
+
+	/**
+	 * Check if a combination of class name and concrete is valid for registration.
+	 * Also return the class constructor if the concrete is either a class name or null (then use the supplied class name).
+	 *
+	 * @param string $class_name The class name to check.
+	 * @param mixed  $concrete   The concrete to check.
+	 *
+	 * @return \ReflectionMethod|null A ReflectionMethod representing the class constructor, if $concrete is null or a class name; null otherwise.
+	 * @throws ContainerException Class has a private constructor, or can't reflect class, or the concrete is invalid.
+	 */
+	private function verify_class_and_concrete( string $class_name, $concrete ) {
+		if ( ! isset( $concrete ) || is_string( $concrete ) && class_exists( $concrete ) ) {
+			try {
+				$class       = $concrete ?? $class_name;
+				$reflector   = new \ReflectionClass( $class );
+				$constructor = $reflector->getConstructor();
+				if ( isset( $constructor ) && ! $constructor->isPublic() ) {
+					throw new ContainerException( "AbstractServiceProvider::add_with_auto_arguments: constructor of class '$class' isn't public, instances can't be created." );
+				}
+				return $constructor;
+			} catch ( \ReflectionException $ex ) {
+				throw new ContainerException( "AbstractServiceProvider::add_with_auto_arguments: error when reflecting class '$class': {$ex->getMessage()}" );
+			}
+		} elseif ( ! is_object( $concrete ) && ! is_callable( $concrete ) && ! function_exists( $concrete ) ) {
```

Comment:
The `! is_callable( $concrete )` also covers the `function_exists` check since a function that does not exist cannot be called.